srk/src/newtonInterpolant.ml

```diff
@@ -0,0 +1,457 @@
+open Syntax
+open BatPervasives
+open Interpolant 
+
+include Log.Make(struct let name = "srk.newtonInterpolant" end)
+
+module NewtonBackwards : Interpolant = functor (C: sig 
+        type t 
+        val context : t context 
+    end)
+    (V: sig
+        type t
+        val show : t -> string
+        val typ : t -> [ `TyInt | `TyReal ]
+        val compare : t -> t -> int
+        val symbol_of : t -> symbol
+        val of_symbol : symbol -> t option
+        val is_global : t -> bool
+    end)
+    (T: sig 
+        type t
+        type var = V.t
+        val equal : t -> t -> bool
+        val compare : t -> t -> int
+
+        (** Guarded parallel assignment *)
+        val construct : C.t formula -> (var * C.t arith_term) list -> t
+
+        val create : C.t formula -> (var * C.t arith_term) BatEnum.t -> t 
+        (** [exists ex tr] removes the variables that do not satisfy the predicate
+            [ex] from the footprint of a transition.  For example, projecting a
+            variable [x] out of a transition [tr] is logically equivalent to
+            [(exists x. tr) && x' = x]. *)
+        val exists : (var -> bool) -> t -> t
+
+        (** (Needed by interpolation) *)
+        val destruct_and : (C.t context) -> (C.t formula) -> (C.t formula) list
+
+        (** Enumerate the variables and values assigned in a transition. *)
+        val transform : t -> (var * C.t arith_term) BatEnum.t
+
+        val defines : t -> var list
+
+        (** Variables used by a a transition, including non-Skolem symbols in both the guard and the transform. *)
+        val uses : t -> var list
+
+        (** The condition under which a transition may be executed. *)
+        val guard : t -> C.t formula
+
+        val state_vocabulary : t -> (var * Syntax.symbol) list 
+  end) -> struct
+  
+    let srk = C.context
+    module M = BatMap.Make(V)
+
+    (* The helper functions below implement the live-variables analysis described in:
```

Comment:
I'm skeptical that this improves interpolant quality.  Can you construct an example where it does?